### PR TITLE
Added ESP8266 MQTT support

### PR DESCRIPTION
**Pull request by debsahu**
Use BearSSL for ESP8266 MQTT secure connection
Address https://github.com/GoogleCloudPlatform/google-cloud-iot-arduino/issues/27

### CODE REVIEW

**Comment by debsahu** [comment]
Update your arduino-esp8266 core on Arduino IDE. Preferably use git version, v2.4.2 should also work!

**Comment by debsahu** [comment]
Please test if it works before merging.

**Comment by debsahu** [comment]
CreateJwt is causing crashes. I tried adding `ESP.wdtDisable();` before CreateJwt is called and `ESP.wdtEnable(1000);` afterwards as suggested by @manojvellala and no help. Still crashes.

**Comment by Gior80** [comment]
@debsahu , which is your output from debug? Can u share crash information ?

**Comment by gguuss** [comment]
@debsahu try ESP.edtEnable(0) - it's also helpful to call delay(10) to let the RTOS feed the WDT.

**Comment by debsahu** [comment]
```
0x40203c62: NN_AddDigitMult(unsigned int*, unsigned int*, unsigned int, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\nn.cpp line 699
0x40203c62: NN_AddDigitMult(unsigned int*, unsigned int*, unsigned int, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\nn.cpp line 699
0x40203d0c: NN_Mult(unsigned int*, unsigned int*, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\nn.cpp line 175
0x40204570: omega_mul(unsigned int*, unsigned int*, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\secp256r1.cpp line 480
0x40203daf: NN_ModMultOpt(unsigned int*, unsigned int*, unsigned int*, unsigned int*, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\nn.cpp line 456
0x402036ae: ecc_m_dbl_projective(point*, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 566
0x40203800: ecc_win_mul(point*, unsigned int*, point*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 626
0x40204570: omega_mul(unsigned int*, unsigned int*, unsigned int*, unsigned char) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\secp256r1.cpp line 480
0x4020397d: ecc_gen_private_key(unsigned int*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 697
0x402038d0: ecc_win_mul_base(point*, unsigned int*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 660
0x402039fc: ecdsa_sign(unsigned char*, unsigned int*, unsigned int*, unsigned int*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecdsa.cpp line 222
0x4020347c: ecc_add(point*, point*, point*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 254
0x40202f33: ecc_get_order(unsigned int*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\crypto\ecc.cpp line 224
0x40204c72: CreateJwt(String, long long, unsigned int*, int) (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266/WString.h line 266
0x4020559c: std::_Vector_base   >::_M_deallocate(unsigned char*, unsigned int) at c:\users\debsahu\appdata\local\arduino15\packages\esp8266\tools\xtensa-lx106-elf-gcc\1.20.0-26-gb404fb9-2\xtensa-lx106-elf\include\c++\4.8.2\bits/stl_vector.h line 175
0x401008a5: realloc (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc\umm_malloc.c line 1503
0x40208bc9: String::copy(char const*, unsigned int) (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\WString.cpp line 175
0x40208c16: String::String(char const*) (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\WString.cpp line 36
0x40202918: CloudIoTCoreDevice::createJWT(long long) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreDevice.cpp line 45
0x40224748: br_x509_decoder_run at src/x509/x509_decoder.c line 717
0x40205d14: std::vector   >::_M_emplace_back_aux (br_x509_certificate const&) at c:\users\debsahu\appdata\local\arduino15\packages\esp8266\tools\xtensa-lx106-elf-gcc\1.20.0-26-gb404fb9-2\xtensa-lx106-elf\include\c++\4.8.2\bits/vector.tcc line 436
0x40202cd1: CloudIoTCoreMQTTClient::getJWT() at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreMQTTClient.cpp line 104
0x40100540: _umm_free (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc\umm_malloc.c line 1299
0x40100a0c: free (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc\umm_malloc.c line 1759
0x402075a3: BearSSL::WiFiClientSecure::connected() (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\libraries\ESP8266WiFi\src\WiFiClientSecureBearSSL.cpp line 253
0x40202dbd: CloudIoTCoreMQTTClient::mqttConnect() at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreMQTTClient.cpp line 146
0x401008a5: realloc (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc\umm_malloc.c line 1503
0x402075a3: BearSSL::WiFiClientSecure::connected() (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\libraries\ESP8266WiFi\src\WiFiClientSecureBearSSL.cpp line 253
0x402097e4: PubSubClient::connected() at C:\Users\debsahu\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp line 606
0x40202d6c: CloudIoTCoreMQTTClient::connected() at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreMQTTClient.cpp line 128
0x40202e92: CloudIoTCoreMQTTClient::loop() at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreMQTTClient.cpp line 180
0x40202744: loop() at C:\Users\debsahu\Dropbox\PIO\GoogleCloudMQTTESP8266\Arduino\GoogleCloudMQTTESP8266/GoogleCloudMQTTESP8266.ino line 102
0x40100a0c: free (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\umm_malloc\umm_malloc.c line 1759
0x402059e4: BearSSL::X509List::~X509List() (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\libraries\ESP8266WiFi\src\BearSSLHelpers.cpp line 783
0x40202c44: CloudIoTCoreMQTTClient::connectSecure(char const*) at C:\Users\debsahu\Documents\Arduino\libraries\google-cloud-iot-arduino\src\CloudIoTCoreMQTTClient.cpp line 92
0x402088e0: Print::println(char const*) (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\Print.cpp line 190
0x402026e8: setup() at C:\Users\debsahu\Dropbox\PIO\GoogleCloudMQTTESP8266\Arduino\GoogleCloudMQTTESP8266/GoogleCloudMQTTESP8266.ino line 99
0x402090b4: loop_wrapper() (C:\Program Files at x86)\Arduino\hardware\esp8266com\esp8266\cores\esp8266\core_esp8266_main.cpp line 125
```

File: examples/Esp8266-mqtt/ciotc_config.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+// Wifi newtork details.
+const char *ssid = "SSID";
+const char *password = "PASSWORD";
+
+// Cloud iot details.
+const char *project_id = "project-id";
+const char *location = "us-central1";
+const char *registry_id = "my-registry";
+const char *device_id = "my-esp32-device";
```
Comment (gguuss) [inline examples/Esp8266-mqtt/ciotc_config.h]: rename to `my-esp8266-device` to clarify this is not for the ESP32

File: examples/Esp8266-mqtt/ciotc_config.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+// To get the private key run (where private-key.pem is the ec private key
+// used to create the certificate uploaded to google cloud iot):
+// openssl ec -in <private-key.pem> -noout -text
+// and copy priv: part.
+// The key length should be exactly the same as the key length bellow (32 pairs
+// of hex digits). If it's bigger and it starts with "00:" delete the "00:". If
+// it's smaller add "00:" to the start. If it's too big or too small something
+// is probably wrong with your key.
+const char *private_key_str =
+"6e:b8:17:35:c7:fc:6b:d7:a9:cb:cb:49:7f:a0:67:"
```
Comment (gguuss) [inline examples/Esp8266-mqtt/ciotc_config.h]: nit. prepend four-spaces as this is a continuation from the previous line

**Comment by Gior80** [comment]
Download esp32 crash analyzer

File: examples/Esp8266-mqtt/ciotc_config.h (hunk 1 above)
Comment (debsahu) [reply examples/Esp8266-mqtt/ciotc_config.h]: done.

File: examples/Esp8266-mqtt/ciotc_config.h (hunk 2 above)
Comment (debsahu) [reply examples/Esp8266-mqtt/ciotc_config.h]: done.

**Comment by gguuss** [comment]
OK, I'm now able to repro the WDT reset, as far as I can tell, it's happening outside of the JWT calculation now.

**Comment by debsahu** [comment]
Esp8266 specific, WiFiClientSecureAxTLS doesn't work with PubSubClient.

**Comment by gguuss** [comment]
@debsahu Did this ever work for you or did you send the PR to debug?

File: src/CloudIoTCoreMQTTClient.cpp [hunk 3]
```diff
@@ -23,19 +23,33 @@ void callback(char *topic, uint8_t *payload, unsigned int length) {
...
 }
+#else
+CloudIoTCoreMQTTClient::CloudIoTCoreMQTTClient(CloudIoTCoreDevice *_device,
+    WiFiClientSecure *_client, PubSubClient *_mqttClient) {
+  this->device = _device;
+  this->client = _client;
+  this->mqttClient = _mqttClient;
+  this->mqttClient->setClient(*(this->client));
+  this->mqttClient->setStream(buffer);
+
```
Comment (gguuss) [inline src/CloudIoTCoreMQTTClient.cpp]: missing closing }

**Comment by gguuss** [comment]
I just noticed I had left a bug in the library, on line 96ish of CloudIoTCoreDevice.cpp, there should be:

```
    void CloudIoTCoreDevice::fillPrivateKey() {                                                                                                                                   
    priv_key[8] = 0;
    ...
```
Or the JWT will be invalid.

**Comment by gguuss** [comment]
Oof, I [created a separate PR](https://github.com/GoogleCloudPlatform/google-cloud-iot-arduino/pull/57) that demonstrates using lwMQTT from the ESP8266 as a standalone example (no CloudIoTCoreMQTTDevice).

**Pull request closed without merging**